### PR TITLE
feat: add claim reward flow

### LINKED ISSUE #868: Claim reward flow
## Claim reward flow

Implement the design from https://www.figma.com/design/fh0OAvj4AS08kHoSxu3DkE/%F0%9F%9A%80-Starknet-Quest?node-id=5850-398&t=c7GUSczdc1oYxVIF-1
into the "Earn" page.

It currently looks like that:
image

Use mock values (template values) for the modal. We'll implement the backend later.

## PROPOSED TODO

- [ ] Add the "Claim all button" :
image

- [ ] Add the claim modal:
image

- [ ] Show the popup (reuse the component we already wrote):
image

- [ ] Make the interactions working: clicking the button -> show the modal -> confirm -> success popup

### PR DESCRIPTION
**Pull request by joeperpetua**
# Pull Request type

Added flow for claim reward in Earn page

Please add the labels corresponding to the type of changes your PR introduces:

- Feature

Resolves: #868

## Other information
Used the Modal component to build both the claim modal and success popup, as the Popup component lacks some styles customization, making it not that great for mobile viewports.

## Summary by CodeRabbit

- **New Features**
	- Introduced `ClaimModal` for users to claim rewards with a responsive design.
	- Added `SuccessModal` to display a confirmation message after claiming rewards.
	- Launched `TokenIcon` for displaying token images with customizable dimensions.

- **Enhancements**
	- Improved `defiTable` functionality with state management for modals related to claims.
	- Enhanced user interaction with clickable areas to trigger modals for claiming rewards.
	- Updated `AppIcon` component for better prop handling and styling options.

### CODE REVIEW

File: components/discover/tokenIcon.tsx [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    USDC: "/icons/usdc.svg",
+    ETH: "/icons/eth.svg",
+    LORDS: "/icons/lords.webp",
+    SITH: "/icons/sith.png",
+    STRK: "/icons/strk.webp",
+  };
+
+  const getTokenIcon = useCallback((tokenName: string) => {
+    return tokenIconMap[tokenName.toUpperCase()] || "/icons/usdc.svg";
+  }, []);
```
[the remark this answers is not in the document]
Comment (joeperpetua) [reply components/discover/tokenIcon.tsx:29]: No unknown icon available to use.  
Defaulting to USDC as implemented in [`TokenSymbol`](https://github.com/lfglabs-dev/starknet.quest/blob/testnet/components/quest-boost/TokenSymbol.tsx)

File: components/discover/claimModal.tsx [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+  <div className="flex flex-row items-center gap-4">
+    <Typography type={TEXT_TYPE.BODY_SMALL}>
+      {currencyValue < 1000 ? currencyValue : `${currencyValue / 1000}K`}
+    </Typography>
+    <Typography type={TEXT_TYPE.BODY_SMALL}>
+      {currencyName}
+    </Typography>
+    <TokenIcon token={currencyName} />
+  </div>
+);
```
[the remark this answers is not in the document]
Comment (joeperpetua) [reply components/discover/claimModal.tsx:56]: I see no need to format to the millions for the supported tokens.

File: components/discover/defiTable.tsx [hunk 3]
```diff
@@ -305,14 +312,33 @@ const DataTable: FunctionComponent<DataTableProps> = ({ data, loading }) => {
...
+            <Typography type={TEXT_TYPE.BODY_MICRO} color="secondary">
+              Find the best opportunities, and earn tokens
+            </Typography>
+          </div>
+          <div
+            onClick={() => setShowClaimModal(true)}
+            className="flex flex-row items-center justify-evenly gap-4 bg-white rounded-lg modified-cursor-pointer h-min lg:mt-2 mt-8 px-4 lg:py-1 py-2">
+            <AppIcon app="starknet" />
+            <Typography type={TEXT_TYPE.BUTTON_LARGE} color="background">
+              Claim all
```
Comment (Marchand-Nicolas) [inline components/discover/defiTable.tsx:329]: The icon is smaller on the design, and the padding is bigger:
image

File: components/discover/claimModal.tsx [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+      aria-label="Reward claim success modal"
+      open={open}
+    >
+      <div className={`${styles.popup} !overflow-y-hidden !rounded-2xl !mt-0 !px-1 !py-1 md:!px-0 md:!py-0`}>
+        <Loading isLoading={loading} loadingType="spinner">
+          <div className={`${styles.popupContent} !px-0 !py-2 md:!px-12 md:!py-6`}>
+            <div className="flex w-full flex-col self-start">
+              <div className="flex w-full lg:flex-row flex-col-reverse lg:items-start items-center justify-between gap-4">
+                <div className="flex flex-col gap-4 lg:text-left text-center">
+                  <Typography type={TEXT_TYPE.BODY_MIDDLE}>
```
Comment (Marchand-Nicolas) [inline components/discover/claimModal.tsx:166]: Could you add a bit of margin here on top of the text:
Screenshot 2024-10-11 at 3 09 13 PM
To look like the design on Figma:
image

**Review by Marchand-Nicolas — CHANGES_REQUESTED** [review]
Looks great, and the code is very clean, well done! Just two little design changes

File: components/discover/defiTable.tsx (hunk 3 above)
Comment (joeperpetua) [reply components/discover/defiTable.tsx:329]: Just touched it a little bit, let me know if it goes well with the design:
```js
<div 
  onClick={() => setShowClaimModal(true)}
  className="flex flex-row items-center justify-evenly gap-4 bg-white rounded-xl modified-cursor-pointer h-min lg:mt-2 mt-8 px-6 py-2.5"
>
  <AppIcon app="starknet" className="w-5 h-5" />
  <Typography type={TEXT_TYPE.BUTTON_LARGE} color="background">
    Claim all
  </Typography>
</div>
```
image

File: components/discover/claimModal.tsx (hunk 4 above)
Comment (joeperpetua) [reply components/discover/claimModal.tsx:166]: And also this one:
image

**Review by Marchand-Nicolas — APPROVED** [review]
Perfect lgtm!

**Comment by Marchand-Nicolas** [comment]
(I will wait for the back implementation in a few days before merging, but you can consider it as merged)

**Pull request merged**